File: backend/src/app/groups/channels.py
"""Per-group notification-channel config API.

A configured group channel REPLACES the global target for that group's
instances (see app.notifications.notifier); deleting it falls back to the
global channel. Managed by superadmins and by admins for groups they are
member of — a non-member gets 404 (invisible, not forbidden). Field shapes,
secret flags and the mask convention live in app.notifications.channel_config.
"""

from __future__ import annotations

import json
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Request, status
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.audit.log import write_audit
from app.auth.deps import require_admin_or_superadmin
from app.crypto.secrets import decrypt, encrypt
from app.db.base import get_session
from app.db.models import Group, GroupChannel, User
from app.net import client_ip
from app.notifications.channel_config import CHANNEL_FIELDS, CHANNELS, MASK
from app.notifications.notifier import _ssrf_block_reason

router = APIRouter(prefix="/groups", tags=["groups"])


class GroupChannelOut(BaseModel):
    channel: str
    config: dict[str, str]  # secrets replaced by MASK
    updated_at: datetime


class GroupChannelPut(BaseModel):
    # Full replace; a secret field equal to MASK keeps the stored value.
    config: dict[str, str]


async def _group_for(session: AsyncSession, group_id: int, user: User) -> Group:
    """Load the group; non-superadmins must be a member (404 otherwise)."""
    group = await session.get(Group, group_id)
    if group is None or (not user.is_superadmin and group_id not in user.group_id_set):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="not found")
    return group


def _masked(channel: str, config: dict[str, str]) -> dict[str, str]:
    fields = {f.name: f for f in CHANNEL_FIELDS[channel]}
    return {
        name: (MASK if fields[name].secret and value else value)
        for name, value in config.items()
        if name in fields
    }


async def _validated_config(
    channel: str, incoming: dict[str, str], existing: dict[str, str]
) -> dict[str, str]:
    """Validate a full-replace payload against the channel's field spec.

    Unknown fields are rejected; a secret sent as MASK keeps the stored value;
    required fields must end up non-empty. Mattermost URLs are SSRF-checked at
    save time (the sender re-checks at send time as defense in depth).
    """
    fields = {f.name: f for f in CHANNEL_FIELDS[channel]}
    unknown = sorted(set(incoming) - set(fields))
    if unknown:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT, detail=f"unknown fields: {unknown}"
        )
    config: dict[str, str] = {}
    for name, field in fields.items():
        value = str(incoming.get(name, "") or "").strip()
        if field.secret and value == MASK:
            value = existing.get(name, "")
        if field.required and not value:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
                detail=f"field '{name}' is required",
            )
        if value and field.options and value not in field.options:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
                detail=f"field '{name}' must be one of {list(field.options)}",
            )
        config[name] = value
    if channel == "email" and config.get("smtp_port"):
        try:
            port = int(config["smtp_port"])
        except ValueError:
            port = 0
        if not 1 <= port <= 65535:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_CONTENT, detail="invalid smtp_port"
            )
    if channel == "mattermost":
        reason = await _ssrf_block_reason(config["url"])
        if reason:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_CONTENT, detail=f"URL rejected: {reason}"
            )
    return config


def _require_channel(channel: str) -> str:
    if channel not in CHANNELS:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="unknown channel")
    return channel


async def _get_row(session: AsyncSession, group_id: int, channel: str) -> GroupChannel | None:
    return (
        await session.execute(
            select(GroupChannel).where(
                GroupChannel.group_id == group_id, GroupChannel.channel == channel
            )
        )
    ).scalar_one_or_none()


@router.get("/{group_id}/channels", response_model=list[GroupChannelOut])
async def list_group_channels(
    group_id: int,
    session: AsyncSession = Depends(get_session),
    user: User = Depends(require_admin_or_superadmin),
) -> list[GroupChannelOut]:
    await _group_for(session, group_id, user)
    rows = (
        (
            await session.execute(
                select(GroupChannel)
                .where(GroupChannel.group_id == group_id)
                .order_by(GroupChannel.channel)
            )
        )
        .scalars()
        .all()
    )
    return [
        GroupChannelOut(
            channel=r.channel,
            config=_masked(r.channel, json.loads(decrypt(r.config_enc))),
            updated_at=r.updated_at,
        )
        for r in rows
    ]


@router.put("/{group_id}/channels/{channel}", response_model=GroupChannelOut)
async def set_group_channel(
    group_id: int,
    channel: str,
    payload: GroupChannelPut,
    request: Request,
    session: AsyncSession = Depends(get_session),
    user: User = Depends(require_admin_or_superadmin),
) -> GroupChannelOut:
    _require_channel(channel)
    await _group_for(session, group_id, user)
    row = await _get_row(session, group_id, channel)
    existing = json.loads(decrypt(row.config_enc)) if row is not None else {}
    config = await _validated_config(channel, payload.config, existing)
    blob = encrypt(json.dumps(config))
    if row is None:
        row = GroupChannel(group_id=group_id, channel=channel, config_enc=blob)
        session.add(row)
    else:
        row.config_enc = blob
    await write_audit(
        session,
        action="group.channel.set",
        result="ok",
        user_id=user.id,
        target_type="group",
        target_id=str(group_id),
        source_ip=client_ip(request),
        detail={"channel": channel, "fields": sorted(k for k, v in config.items() if v)},
    )
    await session.commit()
    await session.refresh(row)
    return GroupChannelOut(
        channel=channel, config=_masked(channel, config), updated_at=row.updated_at
    )


@router.delete("/{group_id}/channels/{channel}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_group_channel(
    group_id: int,
    channel: str,
    request: Request,
    session: AsyncSession = Depends(get_session),
    user: User = Depends(require_admin_or_superadmin),
) -> None:
    _require_channel(channel)
    await _group_for(session, group_id, user)
    row = await _get_row(session, group_id, channel)
    if row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="not configured")
    await session.delete(row)
    await write_audit(
        session,
        action="group.channel.delete",
        result="ok",
        user_id=user.id,
        target_type="group",
        target_id=str(group_id),
        source_ip=client_ip(request),
        detail={"channel": channel},
    )
    await session.commit()

File: orbit/lib/orbit/groups/channels.ex
defmodule Orbit.Groups.Channels do
  @moduledoc """
  Per-group notification-channel config (groups/channels.py port). One
  fernet-encrypted JSON blob per (group, channel); a configured row REPLACES
  the global channel target for that group's instances (Orbit.Notifier reads
  it at dispatch), absence = global fallback.

  Field shapes, secret flags and the mask convention mirror
  notifications/channel_config.py — the UI shows `@mask` instead of secret
  values, and accepts it back as "keep the stored value".
  """

  require Logger

  @mask "••••••"
  @channels ~w(mattermost telegram email)

  # {json key, secret?, required?, options | nil}
  @fields %{
    "mattermost" => [{"url", true, true, nil}],
    "telegram" => [
      {"token", true, true, nil},
      {"chat_id", false, true, nil}
    ],
    "email" => [
      {"smtp_host", false, true, nil},
      {"smtp_port", false, false, nil},
      {"security", false, false, ~w(starttls ssl none)},
      {"from", false, true, nil},
      {"to", false, true, nil},
      {"username", false, false, nil},
      {"password", true, false, nil}
    ]
  }

  def channels, do: @channels
  def mask, do: @mask

  @doc "Field spec for the UI: [%{name, label?, secret, required}]."
  def fields(channel) do
    for {name, secret, required, _options} <- @fields[channel] || [] do
      %{name: name, secret: secret, required: required}
    end
  end

  @doc "Configured channels for a group: %{channel => masked config map}."
  def list(group_id) do
    rows =
      Orbit.Repo.query!(
        "SELECT channel, config_enc FROM group_channels WHERE group_id = ? ORDER BY channel",
        [group_id]
      ).rows

    Map.new(rows, fn [channel, config_enc] ->
      config = config_enc |> Orbit.Crypto.decrypt() |> Jason.decode!()
      {channel, masked(channel, config)}
    end)
  rescue
    e ->
      Logger.warning("group_channels.list_failed group_id=#{group_id} #{Exception.message(e)}")
      %{}
  catch
    kind, reason ->
      Logger.warning("group_channels.list_failed group_id=#{group_id} #{kind} #{inspect(reason)}")
      %{}
  end

  @doc """
  Validate a full-replace payload against the channel's field spec (pure).
  Unknown fields are rejected; a secret sent as the mask keeps the stored
  value; required fields must end up non-empty. Mattermost URLs are
  SSRF-checked at save time (the sender re-checks at send time).
  Returns `{:ok, config} | {:error, message}`.
  """
  def validate(channel, incoming, existing, opts \\ []) do
    fields = @fields[channel] || []
    known = MapSet.new(fields, fn {name, _, _, _} -> name end)
    unknown = incoming |> Map.keys() |> Enum.reject(&MapSet.member?(known, &1)) |> Enum.sort()

    with [] <- unknown,
         {:ok, config} <- build_config(fields, incoming, existing),
         :ok <- check_port(channel, config),
         :ok <- check_ssrf(channel, config, opts) do
      {:ok, config}
    else
      [_ | _] = unknown -> {:error, "unknown fields: #{Enum.join(unknown, ", ")}"}
      {:error, msg} -> {:error, msg}
    end
  end

  defp build_config(fields, incoming, existing) do
    Enum.reduce_while(fields, {:ok, %{}}, fn {name, secret, required, options}, {:ok, acc} ->
      value = incoming |> Map.get(name, "") |> to_string() |> String.trim()
      value = if secret and value == @mask, do: Map.get(existing, name, ""), else: value

      cond do
        required and value == "" ->
          {:halt, {:error, "field '#{name}' is required"}}

        value != "" and options != nil and value not in options ->
          {:halt, {:error, "field '#{name}' must be one of #{Enum.join(options, ", ")}"}}

        true ->
          {:cont, {:ok, Map.put(acc, name, value)}}
      end
    end)
  end

  defp check_port("email", %{"smtp_port" => port}) when port != "" do
    case Integer.parse(port) do
      {n, ""} when n >= 1 and n <= 65_535 -> :ok
      _ -> {:error, "invalid smtp_port"}
    end
  end

  defp check_port(_channel, _config), do: :ok

  defp check_ssrf("mattermost", %{"url" => url}, opts) do
    ssrf_check = Keyword.get(opts, :ssrf_check, &Orbit.Notifier.ssrf_block_reason/1)

    case ssrf_check.(url) do
      nil -> :ok
      reason -> {:error, "URL rejected: #{reason}"}
    end
  end

  defp check_ssrf(_channel, _config, _opts), do: :ok

  @doc """
  Decrypted config for one (group, channel): `{:ok, config}` — `%{}` when the
  row is absent — or `:error` when it could not be read.

  Deliberately NOT collapsing the failure into `%{}`. `upsert/5` feeds this to
  `validate/4` so that a masked secret submitted unchanged keeps its stored
  value; an empty map there means "the operator cleared it", so a read failure
  would have silently wiped the group's stored credentials on the next save.
  A pool checkout exits rather than raising, so the old rescue did not even
  cover the likeliest failure.
  """
  def existing_config(group_id, channel) do
    case Orbit.Repo.query!(
           "SELECT config_enc FROM group_channels WHERE group_id = ? AND channel = ?",
           [group_id, channel]
         ).rows do
      [[config_enc]] -> {:ok, config_enc |> Orbit.Crypto.decrypt() |> Jason.decode!()}
      [] -> {:ok, %{}}
    end
  rescue
    e ->
      Logger.warning("group_channels.existing_config_failed group_id=#{group_id} #{inspect(e)}")
      :error
  catch
    kind, reason ->
      Logger.warning(
        "group_channels.existing_config_failed group_id=#{group_id} #{kind} #{inspect(reason)}"
      )

      :error
  end

  @doc "Validate + persist one channel config. `{:ok, masked} | {:error, msg}`."
  def upsert(group_id, channel, incoming, user, opts \\ []) do
    with true <- channel in @channels || {:error, "unknown channel"},
         {:ok, existing} <- existing_config(group_id, channel),
         {:ok, config} <- validate(channel, incoming, existing, opts) do
      blob = config |> Jason.encode!() |> Orbit.Crypto.encrypt()

      Orbit.Repo.query!(
        "INSERT INTO group_channels (group_id, channel, config_enc, created_at, updated_at) " <>
          "VALUES (?, ?, ?, NOW(), NOW()) " <>
          "ON DUPLICATE KEY UPDATE config_enc = VALUES(config_enc), updated_at = NOW()",
        [group_id, channel, blob]
      )

      Orbit.Audit.write(
        action: "group.channel.set",
        result: "ok",
        user_id: user.id,
        target_type: "group",
        target_id: group_id,
        detail: %{
          "channel" => channel,
          "fields" =>
            config
            |> Enum.filter(fn {_k, v} -> v != "" end)
            |> Enum.map(&elem(&1, 0))
            |> Enum.sort()
        }
      )

      {:ok, masked(channel, config)}
    else
      # Refuse rather than save against an unknown baseline — writing here
      # would blank whatever masked secrets the form did not resend. The
      # caller's else only knows nil and {:error, msg}, so translate.
      :error ->
        {:error, "could not read the current configuration — nothing was saved"}

      other ->
        other
    end
  end

  @doc "Remove one channel override (falls back to the global target)."
  def delete(group_id, channel, user) do
    %{num_rows: n} =
      Orbit.Repo.query!(
        "DELETE FROM group_channels WHERE group_id = ? AND channel = ?",
        [group_id, channel]
      )

    if n > 0 do
      Orbit.Audit.write(
        action: "group.channel.delete",
        result: "ok",
        user_id: user.id,
        target_type: "group",
        target_id: group_id,
        detail: %{"channel" => channel}
      )
    end

    :ok
  end

  @doc "Secrets replaced by the mask; unknown stored keys dropped."
  def masked(channel, config) do
    for {name, secret, _required, _options} <- @fields[channel] || [],
        Map.has_key?(config, name),
        into: %{} do
      value = config[name]
      {name, if(secret and value not in [nil, ""], do: @mask, else: value)}
    end
  end
end
